clients/drcachesim/tests/scheduler_unit_tests.cpp

```diff
@@ -5612,6 +5612,114 @@ test_unscheduled_initially_roi()
 #endif
 }
 
+static void
+test_unscheduled_initially_rebalance()
+{
+    // Tests i#7318 where on a rebalance a too-large runqueue has nothing
+    // but blocked inputs. That's easiest to make happen at the init-time
+    // rebalance where we create a bunch of starts-unscheduled (but not infinite)
+    // inputs.
+    std::cerr << "\n----------------\nTesting initially-unscheduled init rebalance\n";
+    static constexpr int NUM_OUTPUTS = 3;
+    static constexpr int NUM_INPUTS = 5;
+    static constexpr int BLOCK_LATENCY = 100;
+    static constexpr double BLOCK_SCALE = 1. / (BLOCK_LATENCY);
+    static constexpr uint64_t BLOCK_TIME_MAX = 500;
+    static constexpr int MIGRATION_THRESHOLD = 0;
+    static constexpr memref_tid_t TID_BASE = 100;
+    std::vector<trace_entry_t> refs[NUM_INPUTS];
+    for (int i = 0; i < NUM_INPUTS; ++i) {
+        if (i == NUM_OUTPUTS - 1) {
+            // Just one input is runnable.
+            refs[i] = {
+                make_thread(TID_BASE + i),
+                make_pid(1),
+                make_version(TRACE_ENTRY_VERSION),
+                // Runs last by timestamp.
+                make_timestamp(3001),
+                make_marker(TRACE_MARKER_TYPE_CPU_ID, 0),
+                make_instr(/*pc=*/200),
+                make_timestamp(3002),
+                make_marker(TRACE_MARKER_TYPE_CPU_ID, 0),
+                // Makes a long-latency blocking syscall, testing whether
+                // the other threads are really unscheduled.
+                make_marker(TRACE_MARKER_TYPE_SYSCALL, 999),
+                make_marker(TRACE_MARKER_TYPE_MAYBE_BLOCKING_SYSCALL, 0),
+                make_timestamp(7002),
+                make_instr(/*pc=*/201),
+                make_exit(TID_BASE + i),
+            };
+        } else {
+            // The rest start unscheduled.
+            refs[i] = {
+                make_thread(TID_BASE + i),
+                make_pid(1),
+                make_version(TRACE_ENTRY_VERSION),
+                // These have the earliest timestamp and would start.
+                make_timestamp(1001 + i),
+                make_marker(TRACE_MARKER_TYPE_CPU_ID, 0),
+                // They start out unscheduled though.
+                make_marker(TRACE_MARKER_TYPE_SYSCALL_UNSCHEDULE, 0),
+                make_timestamp(4202),
+                make_instr(/*pc=*/102),
+                make_exit(TID_BASE + i),
+            };
+        }
+    }
+    std::vector<scheduler_t::input_reader_t> readers;
+    for (int i = 0; i < NUM_INPUTS; ++i) {
+        readers.emplace_back(std::unique_ptr<mock_reader_t>(new mock_reader_t(refs[i])),
+                             std::unique_ptr<mock_reader_t>(new mock_reader_t()),
+                             TID_BASE + i);
+    }
+    // We need the initial runqueue assignment to be unbalanced.
+    // We achieve that by using input bindings.
+    // This relies on knowing the scheduler takes the 1st binding if there
```

Comment:
Should we document this at input_thread_info_t.output_binding? Or is it just an implementation artifact that the first one gets picked always?

Does the scheduler always take the 1st binding? Why do we allow the user to specify a set then?

Reply:
No this is not visible and should not be documented: this test is relying on an internal implementation detail prior to the rebalance (which will move things around to other bindings before actually running anything).